File: apps/cli/src/commands/mcp.ts
/**
 * `frick mcp` — stdio MCP server for agent access to documented Frick runtime
 * surfaces. Default mode is read-only. `--print-config` emits a JSON client
 * config instead of starting stdio transport.
 */
import { createMcpClientConfig, runFrickMcpStdio, type FrickMcpOptions } from "@frick/mcp";
import { requireBoolean, requireString, type ParsedArgs } from "../argv.js";
import { CliUsageError, EXIT_OK } from "../errors.js";
import { emit, type OutputOptions } from "../output.js";

const DEFAULT_FRICK_ENDPOINT = "http://127.0.0.1:4099";

export async function mcpCommand(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const options = readOptions(parsed);
  if (requireBoolean(parsed.flags, "print-config")) {
    emit(createMcpClientConfig(options), out);
    return EXIT_OK;
  }

  runFrickMcpStdio(options, {
    stdin: process.stdin,
    stdout: process.stdout,
    stderr: process.stderr,
  });
  await waitForever();
  return EXIT_OK;
}

function readOptions(parsed: ParsedArgs): FrickMcpOptions {
  const endpoint = requireString(parsed.flags, "endpoint") ?? process.env.FRICK_ENDPOINT ?? DEFAULT_FRICK_ENDPOINT;
  validateEndpoint(endpoint);
  const token = requireString(parsed.flags, "token");
  const tenantId = requireString(parsed.flags, "tenant");
  const userId = requireString(parsed.flags, "user");

  return {
    endpoint,
    ...(token ? { token } : {}),
    ...(tenantId ? { tenantId } : {}),
    ...(userId ? { userId } : {}),
    allowWrites: requireBoolean(parsed.flags, "allow-writes") && !requireBoolean(parsed.flags, "readonly"),
  };
}

function validateEndpoint(endpoint: string): void {
  try {
    const url = new URL(endpoint);
    if (url.protocol !== "http:" && url.protocol !== "https:") {
      throw new Error("bad protocol");
    }
  } catch {
    throw new CliUsageError(`--endpoint must be an HTTP(S) URL, got ${JSON.stringify(endpoint)}`);
  }
}

function waitForever(): Promise<never> {
  return new Promise(() => {
    // MCP stdio servers stay alive until the host process closes stdin or kills the subprocess.
  });
}

File: crates/frick-cli/src/commands/mcp.rs
//! `frick mcp` (ported from `apps/cli/src/commands/mcp.ts`).
//!
//! `--print-config` emits `createMcpClientConfig(options)` and exits 0.
//! Otherwise it runs the `frick-mcp` stdio JSON-RPC bridge (FR-253) on the
//! process stdio and blocks until stdin closes. `--readonly` wins over
//! `--allow-writes`.

use crate::argv::ParsedArgs;
use crate::errors::{CliError, EXIT_OK};
use crate::mcp_config::{McpClientOptions, create_mcp_client_config, normalize_endpoint};
use crate::output::Output;

const DEFAULT_FRICK_ENDPOINT: &str = "http://127.0.0.1:4099";

/// `mcpCommand`.
pub async fn mcp_command(parsed: &ParsedArgs, out: &mut Output<'_>) -> Result<i32, CliError> {
    let options = read_options(parsed)?;
    if parsed.flag_truthy("print-config") {
        out.emit(&create_mcp_client_config(&options));
        return Ok(EXIT_OK);
    }
    // Run the stdio bridge over the real frick-mcp server; blocks until stdin
    // closes (the TS `runFrickMcpStdio` likewise runs until the transport ends).
    let mcp_options = frick_mcp::FrickMcpOptions {
        endpoint: Some(options.endpoint.clone()),
        token: options.token.clone(),
        tenant_id: options.tenant_id.clone(),
        user_id: options.user_id.clone(),
        allow_writes: options.allow_writes,
    };
    frick_mcp::run_frick_mcp_stdio(&mcp_options)
        .await
        .map_err(|err| CliError::failure("cli.mcp.io", err.to_string()))?;
    Ok(EXIT_OK)
}

fn read_options(parsed: &ParsedArgs) -> Result<McpClientOptions, CliError> {
    let endpoint_raw = parsed
        .flag_str("endpoint")
        .map(ToString::to_string)
        .or_else(|| std::env::var("FRICK_ENDPOINT").ok())
        .unwrap_or_else(|| DEFAULT_FRICK_ENDPOINT.to_string());
    validate_endpoint(&endpoint_raw)?;
    let endpoint = normalize_endpoint(Some(&endpoint_raw));

    // `--readonly` wins over `--allow-writes`.
    let allow_writes = parsed.flag_truthy("allow-writes") && !parsed.flag_truthy("readonly");

    Ok(McpClientOptions {
        endpoint,
        token: parsed.flag_str("token").map(ToString::to_string),
        tenant_id: parsed.flag_str("tenant").map(ToString::to_string),
        user_id: parsed.flag_str("user").map(ToString::to_string),
        allow_writes,
    })
}

fn validate_endpoint(endpoint: &str) -> Result<(), CliError> {
    let ok = endpoint.starts_with("http://") || endpoint.starts_with("https://");
    if ok {
        Ok(())
    } else {
        Err(CliError::usage(format!(
            "--endpoint must be an HTTP(S) URL, got {}",
            serde_json::to_string(endpoint).unwrap_or_else(|_| format!("\"{endpoint}\""))
        )))
    }
}
